Simplenote/SharedStorageMigrator.swift

```diff
@@ -0,0 +1,120 @@
+import Foundation
+import CoreData
+import AutomatticTracks
+
+@objc
+class SharedStorageMigrator: NSObject {
+    private let storageSettings: StorageSettings
+    private let fileManager: FileManager
+
+    init(storageSettings: StorageSettings = StorageSettings(), fileManager: FileManager = FileManager.default) {
+        self.storageSettings = storageSettings
+        self.fileManager = fileManager
+    }
+
+    private var legacyStorageExists: Bool {
+        fileManager.fileExists(atPath: storageSettings.legacyStorageURL.path)
+    }
+
+    private var sharedStorageExists: Bool {
+        fileManager.fileExists(atPath: storageSettings.sharedStorageURL.path)
+    }
+
+    /// Database Migration
+    /// To be able to share data with app extensions, the CoreData database needs to be migrated to an app group
+    /// Must run before Simperium is setup
+
+    func performMigrationIfNeeded() -> MigrationResult {
+        // Confirm if the app group DB exists
+        guard migrationNeeded else {
+            NSLog("Core Data Migration not required")
+            return .notNeeded
+        }
+
+        return migrateCoreDataToAppGroup()
+    }
+
+    private var migrationNeeded: Bool {
+        return legacyStorageExists && !sharedStorageExists
+    }
+
+    private func migrateCoreDataToAppGroup() -> MigrationResult {
+        NSLog("Database needs migration to app group")
+        NSLog("Beginning database migration from: \(storageSettings.legacyStorageURL.path) to: \(storageSettings.sharedStorageURL.path)")
+
+        do {
+            try disableJournaling()
+            try migrateCoreDataFiles()
+            try attemptCreationOfCoreDataStack()
+            NSLog("Database migration successful!!")
+            return .success
+        } catch {
+            NSLog("Could not migrate database to app group " + error.localizedDescription)
+            CrashLogging.logError(error)
+
+            removeFailedMigrationFilesIfNeeded()
+            return .failed
+        }
+    }
+
+    /// This method disables journaling on the core data database
+    /// Per this doc: https://developer.apple.com/library/archive/qa/qa1809/_index.html
+    /// Core data databases have journaling enabled by default, without disabling the journaling first it is possible some notes may get lost in migration
+    ///
+    private func disableJournaling() throws {
+        guard let mom = NSManagedObjectModel(contentsOf: storageSettings.modelURL) else {
+            throw NSError(domain: "SharedStorageMigrator", code: 100, userInfo: nil)
```

Comment:
This one would be a fatal error, if ever happens. You can also use force unwrap for this one BTW

Reply:
Should it though?  In this case if the MOM isn't loaded the migration fails and the app can continue with launching normally.  I mean.... if it fails here it will probably also fail when the core data manager tries the same thing to launch the app, but my thought here was to allow the migrator to fail and the app continue, so I wouldn't think fatalError() would be the right choice for this MOM

Reply:
If MOM is missing, something is really wrong with the app and continuing won't do any good. I also think we should just crash here.

Reply:
makes sense.  Have changed that.